File: frontend/src/lib/types/outreach.ts
export interface Contact {
  id: string;
  user_id: string;
  business_id: string;
  primary_email: string | null;
  primary_phone: string | null;
  display_name: string;
  pipeline_stage: "lead" | "contacted" | "replied" | "qualified" | "won" | "lost";
  default_automation: "manual" | "semi" | "auto";
  default_sequence_id: string | null;
  unsubscribed_at?: string | null;
  unsubscribe_reason?: string | null;
  created_at: string;
  updated_at: string;
}

export interface UserChannel {
  id: string;
  user_id: string;
  type: "gmail_oauth" | "outlook_oauth" | "smtp" | "whatsapp";
  display_label: string;
  from_email: string;
  oauth_expires_at: string | null;
  oauth_scope: string | null;
  enabled: boolean;
  is_default: boolean;
  last_poll_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface SenderProfile {
  id: string;
  name: string;
  user_id: string;
  company_name: string;
  product_description: string;
  value_prop: string;
  target_buyer_description: string;
  tone: string;
  signature: string;
  physical_address: string;
  // Targeting fields (free text — fed into the AI lead scorer so leads are
  // ranked against this user's specific positioning).
  target_industries: string;
  target_countries: string;
  avoid_countries: string;
  min_deal_size_usd: number | null;
  typical_deal_size_usd: number | null;
  deal_breakers: string;
  competitive_moats: string;
  default_channel_id: string | null;
  catalog_file_name?: string | null;
  catalog_mime_type?: string | null;
  catalog_size_bytes?: number | null;
  catalog_uploaded_at?: string | null;
  updated_at: string;
}

export interface Conversation {
  id: string;
  user_id: string;
  contact_id: string;
  channel_id: string;
  campaign_id: string | null;
  channel_type: string;
  subject: string | null;
  external_thread_id: string | null;
  automation: "manual" | "semi" | "auto";
  status: "active" | "paused" | "closed";
  last_message_at: string | null;
  last_direction: "in" | "out" | null;
  unread: boolean;
  created_at: string;
  updated_at: string;
}

export interface ConversationListRow extends Conversation {
  contact_name: string;
  contact_email: string | null;
  business_id: string;
  business_name: string;
  last_message_snippet: string | null;
  message_count: number;
  last_inbound_sentiment: string | null;
  last_inbound_sentiment_score: number | null;
  last_inbound_sentiment_label: SentimentLevel | null;
  tags: string[];
}

export interface Message {
  id: string;
  conversation_id: string;
  direction: "in" | "out";
  channel_type: string;
  external_id: string | null;
  in_reply_to_external_id: string | null;
  subject: string | null;
  body_text: string | null;
  body_html: string | null;
  ai_generated: boolean;
  ai_model: string | null;
  ai_prompt_version: string | null;
  status: string;
  campaign_id: string | null;
  sequence_step_id: string | null;
  sent_at: string | null;
  received_at: string | null;
  created_at: string;
}

export interface StartConversationResult {
  conversation: Conversation;
  draft_message?: Message;
  draft_error?: string;
}

export interface ConversationDetail {
  conversation: Conversation;
  messages: Message[];
}

// --- Campaigns -------------------------------------------------------

export type CampaignStatus =
  | "draft"
  | "ready"
  | "active"
  | "paused"
  | "completed"
  | "stopped";

export type CampaignContactStatus =
  | "pending"
  | "drafted"
  | "approved"
  | "sent"
  | "replied"
  | "cold"
  | "skipped"
  | "failed";

export interface Campaign {
  id: string;
  user_id: string;
  channel_id: string;
  name: string;
  goal: string;
  status: CampaignStatus;
  positioning_override?: Record<string, unknown> | null;
  sequence_id: string | null;
  market_id: string | null;
  sender_profile_id: string | null;
  on_positive_action: string;
  on_negative_action: string;
  send_pace_per_day: number;
  attach_catalog: boolean;
  start_at: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface CampaignSummary extends Campaign {
  pending_count: number;
  drafted_count: number;
  approved_count: number;
  sent_count: number;
  replied_count: number;
  cold_count: number;
  skipped_count: number;
  failed_count: number;
  total_count: number;
}

export interface CampaignContact {
  campaign_id: string;
  contact_id: string;
  market_id: string | null;
  status: CampaignContactStatus;
  draft_message_id: string | null;
  conversation_id: string | null;
  scheduled_send_at: string | null;
  sent_at: string | null;
  replied_at: string | null;
  skip_reason: string | null;
  added_at: string;
}

export interface CampaignContactRow extends CampaignContact {
  ContactName: string;
  ContactEmail: string | null;
  BusinessName: string;
  DraftSubject: string | null;
  DraftBodyText: string | null;
}

export interface AddContactsResult {
  added: number;
  skipped_overlap: string[];
  skipped_already_in_campaign: number;
}

// --- Bulk add-to-contacts result --------------------------------------

export interface BulkEnsureItem {
  business_id: string;
  name: string;
}

export interface BulkEnsureResult {
  added: BulkEnsureItem[];
  already_existed: BulkEnsureItem[];
  no_email: BulkEnsureItem[];
}

// --- Sequences -------------------------------------------------------

export type SequenceTrigger = "no_reply" | "any_reply" | "positive_reply" | "always";
export type SequenceAction =
  | "send_message"
  | "mark_cold"
  | "notify_user"
  | "advance_stage";

export interface Sequence {
  id: string;
  user_id: string;
  name: string;
  description: string;
  is_template: boolean;
  created_at: string;
  updated_at: string;
}

export interface SequenceStep {
  id?: string;
  sequence_id?: string;
  step_number: number;
  wait_days: number;
  trigger: SequenceTrigger;
  action: SequenceAction;
  prompt_override: string | null;
  auto_send: boolean;
}

export interface SequenceWithSteps extends Sequence {
  steps: SequenceStep[];
  active_runs_count: number;
}

// --- CRM tasks + notes ----------------------------------------------

export interface Task {
  id: string;
  user_id: string;
  contact_id: string | null;
  conversation_id: string | null;
  title: string;
  body: string;
  due_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface Note {
  id: string;
  user_id: string;
  contact_id: string;
  body: string;
  created_at: string;
}

// --- Contact groups ---------------------------------------------------

export interface ContactGroup {
  id: string;
  user_id: string;
  name: string;
  sender_profile_id: string | null;
  member_count: number;
  created_at: string;
  updated_at: string;
}

// --- Email Groups -----------------------------------------------------
// An Email Group is a campaign + its attached follow-up steps, presented as
// one thing. These shapes mirror the backend group facade.

// Sentiment is a continuous score (-1..1) surfaced as one of five levels.
export type SentimentLevel =
  | "very_negative"
  | "negative"
  | "neutral"
  | "positive"
  | "very_positive";

// Status facet keys (status + derived buckets).
export type StatusFacet = "replied" | "no_reply" | "cold" | "advanced";

// EmailFacets is the multi-select filter: OR within a facet, AND across facets.
export interface EmailFacets {
  sentiment: SentimentLevel[];
  tags: string[];
  status: StatusFacet[];
}

export interface GroupSummary extends CampaignSummary {
  contact_group_id: string | null;
  contact_group_name: string;
  brand_id: string | null;
  brand_name: string;
  sender_email: string;
}

export interface GroupDetail extends GroupSummary {
  steps: SequenceStep[];
  facet_counts: Record<string, number>;
  notifications: Task[];
}

export interface GroupEmail {
  contact_id: string;
  contact_name: string;
  contact_email: string | null;
  business_name: string;
  conversation_id: string | null;
  subject: string | null;
  status: CampaignContactStatus;
  last_direction: "in" | "out" | null;
  last_message_at: string | null;
  sentiment: string | null;
  sentiment_score: number | null;
  sentiment_label: SentimentLevel | null;
  tags: string[];
  unread: boolean;
  has_pending_draft: boolean;
  scheduled_send_at: string | null;
  current_step: number | null;
  next_run_at: string | null;
}

export interface FunnelStep {
  step: number;
  label: string;
  sent: number;
  pct: number;
}

export interface UpcomingDay {
  date: string;
  cold: number;
  followup: number;
}

export interface CadenceStats {
  total_contacts: number;
  funnel: FunnelStep[];
  exits: Record<string, number>;
  completion_pct: number;
  upcoming: UpcomingDay[];
}

// ConvTag is a stored intent tag with its provenance (ai|manual).
export interface ConvTag {
  tag: string;
  source: "ai" | "manual";
  confidence: number | null;
}

// PlaybookEntry is one authored per-tag reply instruction for a group.
export interface PlaybookEntry {
  tag: string;
  instruction: string;
}

// AssistantAction is a write the agent proposes (the user confirms it).
export interface AssistantAction {
  type: "edit_drafts" | "add_playbook" | "remove_playbook";
  scope?: "cold" | "reply" | "reply_positive" | "reply_negative" | "all";
  instruction?: string;
  tag?: string;
}

// AssistantMessage is one turn in the group's draft-assistant chat thread.
export interface AssistantMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  proposed_action?: AssistantAction | null;
  status: "sent" | "proposed" | "applied" | "dismissed";
  created_at: string;
}

// BrandLesson is a remembered per-brand reply instruction, matched by tags +
// sentiment so it only applies to the same kind of buyer reply.
export interface BrandLesson {
  id: string;
  instruction: string;
  match_tags: string[];
  match_sentiment: string;
  created_at: string;
}

File: backend-go/internal/domain/outreach.go
package domain

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// --- Contact ------------------------------------------------------------

// Contact is a user's CRM relationship with a business.
// One contact per (user_id, business_id). Survives across markets.
type Contact struct {
	ID                uuid.UUID  `json:"id" db:"id"`
	UserID            uuid.UUID  `json:"user_id" db:"user_id"`
	BusinessID        uuid.UUID  `json:"business_id" db:"business_id"`
	PrimaryEmail      *string    `json:"primary_email" db:"primary_email"`
	PrimaryPhone      *string    `json:"primary_phone" db:"primary_phone"`
	DisplayName       string     `json:"display_name" db:"display_name"`
	PipelineStage     string     `json:"pipeline_stage" db:"pipeline_stage"`
	DefaultAutomation string     `json:"default_automation" db:"default_automation"`
	DefaultSequenceID *uuid.UUID `json:"default_sequence_id" db:"default_sequence_id"`
	UnsubscribedAt    *time.Time `json:"unsubscribed_at,omitempty" db:"unsubscribed_at"`
	UnsubscribeReason *string    `json:"unsubscribe_reason,omitempty" db:"unsubscribe_reason"`
	CreatedAt         time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt         time.Time  `json:"updated_at" db:"updated_at"`
}

// Pipeline stages — denormalized enum as strings for flexibility.
const (
	PipelineLead       = "lead"
	PipelineContacted  = "contacted"
	PipelineReplied    = "replied"
	PipelineQualified  = "qualified"
	PipelineWon        = "won"
	PipelineLost       = "lost"
)

const (
	AutomationManual = "manual"
	AutomationSemi   = "semi"
	AutomationAuto   = "auto"
)

// --- UserChannel --------------------------------------------------------

// UserChannel is one sending/receiving channel configured by a user.
// Tokens are stored encrypted at rest; the struct holds plaintext in memory.
type UserChannel struct {
	ID                       uuid.UUID       `json:"id" db:"id"`
	UserID                   uuid.UUID       `json:"user_id" db:"user_id"`
	Type                     string          `json:"type" db:"type"`
	DisplayLabel             string          `json:"display_label" db:"display_label"`
	FromEmail                string          `json:"from_email" db:"from_email"`
	OAuthAccessTokenCipher   *string         `json:"-" db:"oauth_access_token_encrypted"`
	OAuthRefreshTokenCipher  *string         `json:"-" db:"oauth_refresh_token_encrypted"`
	OAuthExpiresAt           *time.Time      `json:"oauth_expires_at" db:"oauth_expires_at"`
	OAuthScope               *string         `json:"oauth_scope" db:"oauth_scope"`
	ConfigCipher             json.RawMessage `json:"-" db:"config_encrypted"`
	UnsubscribeSecret        []byte          `json:"-" db:"unsubscribe_secret"`
	Enabled                  bool            `json:"enabled" db:"enabled"`
	IsDefault                bool            `json:"is_default" db:"is_default"`
	LastPollAt               *time.Time      `json:"last_poll_at" db:"last_poll_at"`
	CreatedAt                time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt                time.Time       `json:"updated_at" db:"updated_at"`
}

// Channel type constants.
const (
	ChannelTypeGmailOAuth   = "gmail_oauth"
	ChannelTypeOutlookOAuth = "outlook_oauth"
	ChannelTypeSMTP         = "smtp"
	ChannelTypeWhatsApp     = "whatsapp"
)

// SendReady reports whether the channel has the credentials it needs to
// actually send. A channel can be enabled (listed, selectable) yet missing
// its stored credentials — e.g. an SMTP row whose config was never
// persisted, or an OAuth row whose tokens were cleared on disconnect.
// Binding such a channel to a campaign fails only at send time with no
// visible error, so callers must gate on this before binding.
func (uc UserChannel) SendReady() bool {
	switch uc.Type {
	case ChannelTypeSMTP:
		return len(uc.ConfigCipher) > 0
	case ChannelTypeGmailOAuth, ChannelTypeOutlookOAuth:
		return uc.OAuthRefreshTokenCipher != nil && *uc.OAuthRefreshTokenCipher != ""
	default:
		return true
	}
}

// --- SenderProfile ------------------------------------------------------

// SenderProfile is the "what I sell" context used for AI personalization
// AND for AI lead scoring (so leads are ranked against THIS user's ICP,
// not just the market's product). One per user. Campaigns may override
// specific fields.
type SenderProfile struct {
	// ID is the brand's primary key. A user can have many sender profiles
	// (one brand per market). UserID is now just the owner reference.
	ID                     uuid.UUID  `json:"id" db:"id"`
	Name                   string     `json:"name" db:"name"`
	UserID                 uuid.UUID  `json:"user_id" db:"user_id"`
	CompanyName            string     `json:"company_name" db:"company_name"`
	ProductDescription     string     `json:"product_description" db:"product_description"`
	ValueProp              string     `json:"value_prop" db:"value_prop"`
	TargetBuyerDescription string     `json:"target_buyer_description" db:"target_buyer_description"`
	Tone                   string     `json:"tone" db:"tone"`
	Signature              string     `json:"signature" db:"signature"`
	PhysicalAddress        string     `json:"physical_address" db:"physical_address"`
	DefaultChannelID       *uuid.UUID `json:"default_channel_id" db:"default_channel_id"`

	// Targeting fields (added by migration 17). Free-text by design —
	// the AI scorer reads them as natural language so the user can
	// express intent without locking us into rigid taxonomies.
	TargetIndustries    string `json:"target_industries" db:"target_industries"`
	TargetCountries     string `json:"target_countries" db:"target_countries"`
	AvoidCountries      string `json:"avoid_countries" db:"avoid_countries"`
	MinDealSizeUSD      *int64 `json:"min_deal_size_usd" db:"min_deal_size_usd"`
	TypicalDealSizeUSD  *int64 `json:"typical_deal_size_usd" db:"typical_deal_size_usd"`
	DealBreakers        string `json:"deal_breakers" db:"deal_breakers"`
	CompetitiveMoats    string `json:"competitive_moats" db:"competitive_moats"`

	// Catalog metadata. Bytes live in sender_profiles.catalog_data but are
	// NEVER included in the JSON response — fetched separately during send.
	CatalogFileName   *string    `json:"catalog_file_name,omitempty" db:"catalog_file_name"`
	CatalogMimeType   *string    `json:"catalog_mime_type,omitempty" db:"catalog_mime_type"`
	CatalogSizeBytes  *int       `json:"catalog_size_bytes,omitempty" db:"catalog_size_bytes"`
	CatalogUploadedAt *time.Time `json:"catalog_uploaded_at,omitempty" db:"catalog_uploaded_at"`

	UpdatedAt time.Time `json:"updated_at" db:"updated_at"`
}

// --- Conversation -------------------------------------------------------

// Conversation is a threaded message log on a single channel.
type Conversation struct {
	ID               uuid.UUID  `json:"id" db:"id"`
	UserID           uuid.UUID  `json:"user_id" db:"user_id"`
	ContactID        uuid.UUID  `json:"contact_id" db:"contact_id"`
	ChannelID        uuid.UUID  `json:"channel_id" db:"channel_id"`
	CampaignID       *uuid.UUID `json:"campaign_id" db:"campaign_id"`
	ChannelType      string     `json:"channel_type" db:"channel_type"`
	Subject          *string    `json:"subject" db:"subject"`
	ExternalThreadID *string    `json:"external_thread_id" db:"external_thread_id"`
	Automation       string     `json:"automation" db:"automation"`
	Status           string     `json:"status" db:"status"`
	LastMessageAt    *time.Time `json:"last_message_at" db:"last_message_at"`
	LastDirection    *string    `json:"last_direction" db:"last_direction"`
	Unread           bool       `json:"unread" db:"unread"`
	CreatedAt        time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at" db:"updated_at"`
}

const (
	ConversationStatusActive = "active"
	ConversationStatusPaused = "paused"
	ConversationStatusClosed = "closed"

	DirectionOutbound = "out"
	DirectionInbound  = "in"
)

// --- Message ------------------------------------------------------------

// Message is a single message in a conversation, channel-agnostic.
type Message struct {
	ID                    uuid.UUID  `json:"id" db:"id"`
	ConversationID        uuid.UUID  `json:"conversation_id" db:"conversation_id"`
	Direction             string     `json:"direction" db:"direction"`
	ChannelType           string     `json:"channel_type" db:"channel_type"`
	ExternalID            *string    `json:"external_id" db:"external_id"`
	InReplyToExternalID   *string    `json:"in_reply_to_external_id" db:"in_reply_to_external_id"`
	Subject               *string    `json:"subject" db:"subject"`
	BodyText              *string    `json:"body_text" db:"body_text"`
	BodyHTML              *string    `json:"body_html" db:"body_html"`
	AIGenerated           bool       `json:"ai_generated" db:"ai_generated"`
	AIModel               *string    `json:"ai_model" db:"ai_model"`
	AIPromptVersion       *string    `json:"ai_prompt_version" db:"ai_prompt_version"`
	Status                string     `json:"status" db:"status"`
	CampaignID            *uuid.UUID `json:"campaign_id" db:"campaign_id"`
	SequenceStepID        *uuid.UUID `json:"sequence_step_id" db:"sequence_step_id"`
	SentAt                *time.Time `json:"sent_at" db:"sent_at"`
	ReceivedAt            *time.Time `json:"received_at" db:"received_at"`
	CreatedAt             time.Time  `json:"created_at" db:"created_at"`
}

const (
	MessageStatusDraft            = "draft"
	MessageStatusPendingApproval  = "pending_approval"
	MessageStatusQueued           = "queued"
	MessageStatusSent             = "sent"
	MessageStatusBounced          = "bounced"
	MessageStatusOpened           = "opened"
	MessageStatusFailed           = "failed"
)

// --- Combined views for handler responses ------------------------------

// ConversationWithContext is returned in inbox and detail endpoints —
// joins contact info, business name, last message preview.
type ConversationWithContext struct {
	Conversation
	ContactName       string    `json:"contact_name"`
	ContactEmail      *string   `json:"contact_email"`
	BusinessID        uuid.UUID `json:"business_id"`
	BusinessName      string    `json:"business_name"`
	LastMessageSnippet *string  `json:"last_message_snippet"`
	MessageCount       int      `json:"message_count"`
}

// --- Campaign -----------------------------------------------------------

// Campaign is a named batch of outreach with shared positioning.
type Campaign struct {
	ID                   uuid.UUID       `json:"id" db:"id"`
	UserID               uuid.UUID       `json:"user_id" db:"user_id"`
	ChannelID            uuid.UUID       `json:"channel_id" db:"channel_id"`
	Name                 string          `json:"name" db:"name"`
	Goal                 string          `json:"goal" db:"goal"`
	Status               string          `json:"status" db:"status"`
	PositioningOverride  json.RawMessage `json:"positioning_override,omitempty" db:"positioning_override"`
	SequenceID           *uuid.UUID      `json:"sequence_id" db:"sequence_id"`
	// MarketID scopes the campaign (Email Group) to one market; SenderProfileID
	// is the brand it sends as (resolved from that market at creation). Both
	// nullable so legacy campaigns survive — workers fall back to the user's
	// Default brand when SenderProfileID is nil.
	MarketID             *uuid.UUID      `json:"market_id" db:"market_id"`
	ContactGroupID       *uuid.UUID      `json:"contact_group_id" db:"contact_group_id"`
	SenderProfileID      *uuid.UUID      `json:"sender_profile_id" db:"sender_profile_id"`
	// Response-aware follow-up branches: what to do when a recipient replies.
	OnPositiveAction     string          `json:"on_positive_action" db:"on_positive_action"`
	OnNegativeAction     string          `json:"on_negative_action" db:"on_negative_action"`
	SendPacePerDay       int             `json:"send_pace_per_day" db:"send_pace_per_day"`
	AttachCatalog        bool            `json:"attach_catalog" db:"attach_catalog"`
	StartAt              *time.Time      `json:"start_at" db:"start_at"`
	StartedAt            *time.Time      `json:"started_at" db:"started_at"`
	CompletedAt          *time.Time      `json:"completed_at" db:"completed_at"`
	CreatedAt            time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt            time.Time       `json:"updated_at" db:"updated_at"`
}

const (
	CampaignStatusDraft     = "draft"
	CampaignStatusReady     = "ready"
	CampaignStatusActive    = "active"
	CampaignStatusPaused    = "paused"
	CampaignStatusCompleted = "completed"
	CampaignStatusStopped   = "stopped"
)

// CampaignContact is the membership row tying a contact to a campaign with
// its per-row drafting/send state.
type CampaignContact struct {
	CampaignID       uuid.UUID  `json:"campaign_id" db:"campaign_id"`
	ContactID        uuid.UUID  `json:"contact_id" db:"contact_id"`
	MarketID         *uuid.UUID `json:"market_id" db:"market_id"`
	Status           string     `json:"status" db:"status"`
	DraftMessageID   *uuid.UUID `json:"draft_message_id" db:"draft_message_id"`
	ConversationID   *uuid.UUID `json:"conversation_id" db:"conversation_id"`
	ScheduledSendAt  *time.Time `json:"scheduled_send_at" db:"scheduled_send_at"`
	SentAt           *time.Time `json:"sent_at" db:"sent_at"`
	RepliedAt        *time.Time `json:"replied_at" db:"replied_at"`
	SkipReason       *string    `json:"skip_reason" db:"skip_reason"`
	AddedAt          time.Time  `json:"added_at" db:"added_at"`
}

const (
	CampaignContactPending  = "pending"
	CampaignContactDrafted  = "drafted"
	CampaignContactApproved = "approved"
	CampaignContactSent     = "sent"
	CampaignContactReplied  = "replied"
	CampaignContactCold     = "cold"
	CampaignContactSkipped  = "skipped"
	CampaignContactFailed   = "failed"
)

// CampaignSummary is the dashboard-friendly view of a campaign with its
// per-status counts, used by the campaign detail page.
type CampaignSummary struct {
	Campaign
	PendingCount  int `json:"pending_count"`
	DraftedCount  int `json:"drafted_count"`
	ApprovedCount int `json:"approved_count"`
	SentCount     int `json:"sent_count"`
	RepliedCount  int `json:"replied_count"`
	ColdCount     int `json:"cold_count"`
	SkippedCount  int `json:"skipped_count"`
	FailedCount   int `json:"failed_count"`
	TotalCount    int `json:"total_count"`
}

// --- Sequence -----------------------------------------------------------

// Sequence is a follow-up playbook (a list of timed steps).
type Sequence struct {
	ID          uuid.UUID `json:"id" db:"id"`
	UserID      uuid.UUID `json:"user_id" db:"user_id"`
	Name        string    `json:"name" db:"name"`
	Description string    `json:"description" db:"description"`
	IsTemplate  bool      `json:"is_template" db:"is_template"`
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time `json:"updated_at" db:"updated_at"`
}

// SequenceStep is one step in a sequence.
type SequenceStep struct {
	ID             uuid.UUID `json:"id" db:"id"`
	SequenceID     uuid.UUID `json:"sequence_id" db:"sequence_id"`
	StepNumber     int       `json:"step_number" db:"step_number"`
	WaitDays       int       `json:"wait_days" db:"wait_days"`
	Trigger        string    `json:"trigger" db:"trigger"`
	Action         string    `json:"action" db:"action"`
	PromptOverride *string   `json:"prompt_override" db:"prompt_override"`
	AutoSend       bool      `json:"auto_send" db:"auto_send"`
}

const (
	SequenceTriggerNoReply       = "no_reply"
	SequenceTriggerAnyReply      = "any_reply"
	SequenceTriggerPositiveReply = "positive_reply"
	SequenceTriggerAlways        = "always"

	SequenceActionSendMessage  = "send_message"
	SequenceActionMarkCold     = "mark_cold"
	SequenceActionNotifyUser   = "notify_user"
	SequenceActionAdvanceStage = "advance_stage"
)

// Response-branch actions: what a campaign does when a recipient replies.
const (
	OnPositiveAutoDraftReply = "auto_draft_reply" // AI drafts a reply for approval (default)
	OnPositiveNotify         = "notify"           // flag it; user takes over
	OnPositiveAutoSend       = "auto_send"        // AI replies and sends (needs prior human send)
	OnNegativeMarkCold       = "mark_cold"        // stop + mark the contact cold (default)
	OnNegativeNotify         = "notify"           // flag it; user takes over
)

// Inbound sentiment buckets (mirrors the classifier output).
const (
	SentimentPositive = "positive"
	SentimentNeutral  = "neutral"
	SentimentNegative = "negative"
)

// SequenceWithSteps is the API-facing sequence, used in handler responses.
type SequenceWithSteps struct {
	Sequence
	Steps           []SequenceStep `json:"steps"`
	ActiveRunsCount int            `json:"active_runs_count"`
}

// SequenceRun tracks one conversation's progress through a sequence.
type SequenceRun struct {
	ID             uuid.UUID `json:"id" db:"id"`
	SequenceID     uuid.UUID `json:"sequence_id" db:"sequence_id"`
	ConversationID uuid.UUID `json:"conversation_id" db:"conversation_id"`
	CurrentStep    int       `json:"current_step" db:"current_step"`
	NextRunAt      time.Time `json:"next_run_at" db:"next_run_at"`
	Status         string    `json:"status" db:"status"`
	LastError      *string   `json:"last_error" db:"last_error"`
	CreatedAt      time.Time `json:"created_at" db:"created_at"`
	UpdatedAt      time.Time `json:"updated_at" db:"updated_at"`
}

const (
	SequenceRunActive          = "active"
	SequenceRunPaused          = "paused"
	SequenceRunCompleted       = "completed"
	SequenceRunStoppedOnReply  = "stopped_on_reply"
)

// --- Contact groups -----------------------------------------------------

// ContactGroup is a named, user-owned collection of contacts with its own
// brand. Email groups are built from a contact group.
type ContactGroup struct {
	ID              uuid.UUID  `json:"id" db:"id"`
	UserID          uuid.UUID  `json:"user_id" db:"user_id"`
	Name            string     `json:"name" db:"name"`
	SenderProfileID *uuid.UUID `json:"sender_profile_id" db:"sender_profile_id"`
	CreatedAt       time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at" db:"updated_at"`
}

// --- Tasks & notes ------------------------------------------------------

// Task is a CRM to-do, optionally tied to a contact and/or conversation.
type Task struct {
	ID             uuid.UUID  `json:"id" db:"id"`
	UserID         uuid.UUID  `json:"user_id" db:"user_id"`
	ContactID      *uuid.UUID `json:"contact_id" db:"contact_id"`
	ConversationID *uuid.UUID `json:"conversation_id" db:"conversation_id"`
	Title          string     `json:"title" db:"title"`
	Body           string     `json:"body" db:"body"`
	DueAt          *time.Time `json:"due_at" db:"due_at"`
	CompletedAt    *time.Time `json:"completed_at" db:"completed_at"`
	CreatedAt      time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at" db:"updated_at"`
}

// Note is a free-text annotation attached to a contact.
type Note struct {
	ID        uuid.UUID `json:"id" db:"id"`
	UserID    uuid.UUID `json:"user_id" db:"user_id"`
	ContactID uuid.UUID `json:"contact_id" db:"contact_id"`
	Body      string    `json:"body" db:"body"`
	CreatedAt time.Time `json:"created_at" db:"created_at"`
}
